service/rest_version.go

```diff
@@ -180,27 +186,74 @@ func (restapi restAPI) getRecentVersions(w http.ResponseWriter, r *http.Request)
 
 		result.Versions = append(result.Versions, versionInfo)
 	}
+	// Find all builds/tasks corresponding the set of version ids
+	if err = result.populateBuildsAndTasks(versionIds, versionIdx); err != nil {
+		msg := fmt.Sprintf("Error populating builds/tasks for recent versions of project '%v'", projectId)
+		grip.Errorf("%v: %+v", msg, err)
+		gimlet.WriteJSONInternalError(w, responseError{Message: msg})
+	}
 
-	for _, build := range builds {
-		buildInfo := versionBuildInfo{
-			Id:    build.Id,
-			Name:  build.DisplayName,
-			Tasks: make(versionByBuildByTask, len(build.Tasks)),
-		}
-
-		for _, task := range build.Tasks {
-			buildInfo.Tasks[task.DisplayName] = versionStatus{
-				Id:        task.Id,
-				Status:    task.Status,
-				TimeTaken: task.TimeTaken,
-			}
+	// create a page header
+	if nextPageStart != "" {
+		responder := gimlet.NewResponseBuilder()
+		err = responder.SetPages(&gimlet.ResponsePages{
+			Next: &gimlet.Page{
+				Relation:        "next",
+				LimitQueryParam: "limit",
+				KeyQueryParam:   "start",
+				BaseURL:         restapi.GetSettings().ApiUrl,
+				Key:             nextPageStart,
+				Limit:           l,
+			},
+		})
+		if err != nil {
+			msg := "error setting pages"
+			grip.Errorf("%v: %+v", msg, err)
+			gimlet.WriteJSONInternalError(w, responseError{Message: msg})
 		}
+		w.Header().Set("Link", responder.Pages().GetLinks(r.URL.String()))
+	}
+	gimlet.WriteJSON(w, result)
+}
 
-		versionInfo := result.Versions[versionIdx[build.Version]]
-		versionInfo.Builds[build.BuildVariant] = buildInfo
+func (r *recentVersionsContent) populateBuildsAndTasks(versionIds []string, versionIdx map[string]int) error {
+	builds, err := build.Find(
+		build.ByVersions(versionIds).
+			WithFields(build.BuildVariantKey, build.DisplayNameKey, build.TasksKey, build.VersionKey))
+	if err != nil {
+		return errors.Wrap(err, "Error finding recent versions")
+	}
+	tasks, err := task.Find(task.ByVersions(versionIds).
+		WithFields(task.IdKey, task.DisplayNameKey, task.StatusKey, task.TimeTakenKey, task.VersionKey, task.BuildVariantKey))
```

Comment:
can we just write model functions with the projections in the model functions directly so that it's all encapsulated=?